Initial Ewald CPU implementation and tests

jpcoles opened the pull request:
A CPU implementation of periodic boundaries using the Ewald summation method. The main entry point is the function computeGravityEwald(), which takes the same arguments are computeGravity().

sekelle commented on ryoanji/src/ryoanji/nbody/traversal_ewald_cpu.hpp:
```diff
@@ -0,0 +1,600 @@
+        T1* ax, T1* ay, T1* az, T1* ugrav,
+        LocalIndex numParticles,
+        int gravity_switch
+        )
```
We should have this function minus the `j`-loop and implement the loop one level above this function. We'll pass distance from root-center instead of coordinate arrays and return acceleration contribution by value. Then we'll be able to call this function computing the long-range correction also from a GPU kernel.

jpcoles replied:
ok, I'll do this.

sekelle commented on ryoanji/src/ryoanji/nbody/traversal_ewald_cpu.hpp:
```diff
@@ -0,0 +1,600 @@
+                                const T1 dx_target, const T1 dy_target, const T1 dz_target,
+                                const T1* x, const T1* y, const T1* z, const T2* h, const Tm* m, float G, T1* ax, T1* ay,
+                                T1* az, T1* ugrav)
+{
```
If we change the signature here to directly take `targetCenter` and `targetSize`, also for the existing variant of this function, and compute targetCenter and targetSize at the call site, we only need one version of this function to traverse both the center box and replicas.

jpcoles replied:
computeGravityGroupReplica() differs from computeGravityGroup() in more ways than just the value of targetCenter. First, the P2P loop assumes that all particles in the target leaf interact with the particles in the replica leaf. This is different than the normal case where the target must not be its own source. Second, I need to apply {dx,dy,dz}_target to each target particle (in both P2P and M2P). One could pass a new array of target x,y,z values, but then a new function would still be required and would, I feel, overcomplicate the calling code.

jpcoles commented:
I will do a squash rebase into a single commit on all commits related to ewald once it is approved. Please don't do the rebase yet.

sekelle reviewed (dismissed):
From my side, this is ready now for squash & merge.

sekelle committed:
Ewald CPU implementation

scripts/pre-commit: Accept an optional commit number from the command line.
gravity_wrapper.hpp: Update call to computeGravity to pass NULL pointer for ugrav array

sekelle commented:
@jpcoles will need to recreate this commit himself, please nobody merge this yet.
